File: ruby/perfect-numbers/perfect_numbers.rb
# frozen_string_literal: true

class PerfectNumber
  def self.classify(number)
    raise RuntimeError if number.negative?

    number_factors_sum = (1..number / 2).sum { |n| (number % n).zero? ? n : 0 }
    if number_factors_sum < number
      'deficient'
    elsif number_factors_sum > number
      'abundant'
    else
      'perfect'
    end
  end
end

File: python/perfect-numbers/perfect_numbers.py
def classify(number):
    if number <= 0:
        raise ValueError('Number must be positive')

    num_dividers = [x for x in range(1, number // 2 + 1) if number % x == 0]
    sum_dividers = sum(num_dividers)

    if sum_dividers == number:
        return 'perfect'
    elif sum_dividers < number:
        return 'deficient'
    else:
        return 'abundant'
